packages/widget/src/context/wallet.ts

```diff
@@ -54,14 +70,76 @@ export class WalletContextProvider extends BaseComponent {
   @provide({ context: walletContext })
   private walletContext: WalletContext = {};
 
+  @provide({ context: substrateProviderContext })
+  substrateProviderContext: SubstrateProviderContext = {};
+
   @property({ attribute: false, type: Object })
   evmWallet?: EvmWallet;
 
-  connectedCallback(): void {
+  @property({ attribute: false })
+  substrateProviders?: Array<ApiPromise> = [];
+
+  @property({ type: String }) environment?: Environment;
+
+  get defaultSubstrateProviders(): Array<ApiPromise> {
+    const defaultConnections: Array<ApiPromise> = [];
+    const environment = this.environment ?? Environment.TESTNET;
+
+    for (const connection of SUBSTRATE_CHAIN_RPCS[environment]) {
+      const provider = new WsProvider(connection);
+      const api = new ApiPromise({ provider });
+      defaultConnections.push(api);
+    }
+
+    return defaultConnections;
+  }
+
+  /**
+   * Creates a provider map w.r.t parachain ids
+   * @param {Array<ApiPromise>} providers array of {@link ApiPromise}
+   * @returns {Promise<ParachainProviders>}
+   */
+  async createProvidersMap(
+    providers: Array<ApiPromise>
+  ): Promise<ParachainProviders> {
+    const map: ParachainProviders = new Map();
+
+    for (const provider of providers) {
+      try {
+        await provider.isReady;
+        // TODO: use polkadot type augmentation to remove "as 32"
+        const parachainId = await provider.query.parachainInfo.parachainId();
+        const _parachainId = (parachainId as u32).toNumber();
+        map.set(_parachainId, provider);
+      } catch (error) {
+        console.error('user provided substrate provider ignored', error);
+      }
+    }
+
+    return map;
+  }
+
+  async connectedCallback(): Promise<void> {
     super.connectedCallback();
     if (this.evmWallet) {
       this.walletContext.evmWallet = this.evmWallet;
     }
+
+    let substrateProviders;
+    if (this.substrateProviders) {
+      substrateProviders = await this.createProvidersMap(
+        this.substrateProviders
```

Comment:
why not combine with this.defaultSubstrateProviders?

Reply:
defaultSubstrateProviders should not be coupled with the specified providers logics imo. having a separate method is fine and `defaultSubstrateProviders` should always provide `ApiPromise` objects that are hardcoded or from our config

Reply:
I'm not sure I follow logic here.
Let's take for example, I am developer using this widget.
I have better endpoint for one parachain so I've put it into widget property. With current logic, only routes to parachain will work and other will not because default values are ignored. So I have to copy paste default values from widget.

More natural flow would be that I send one for which I have better endpoint and rest of the defaults stays the same.

Or do you envision this to work in some other way?

Reply:
@mpetrunic okay i follow you now, so few questions
1. since defaultProviders are a must have due to the flow you suggested, is it fine to hardcode these on UI/widget?
2. Should we change the interface as well to accept a `Map` rather than array of providers?